test/integration/ws/w_socket_common.dart

```diff
@@ -0,0 +1,293 @@
+// Copyright 2015 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+library w_transport.test.integration.ws.w_socket_common;
+
+import 'dart:async';
+
+import 'package:test/test.dart';
+import 'package:w_transport/w_transport.dart' show WSocket, WSocketException;
+
+/// These are WebSocket integration tests that should work from client or server.
+/// These will not pass if run on their own!
+void run(String usage) {
+  group('WSocket ($usage)', () {
+    WSocket socket;
+    Uri closeUri;
+    Uri echoUri;
+    Uri pingUri;
+
+    setUp(() {
+      closeUri = Uri.parse('ws://localhost:8024/test/ws/close');
+      echoUri = Uri.parse('ws://localhost:8024/test/ws/echo');
+      pingUri = Uri.parse('ws://localhost:8024/test/ws/ping');
+    });
+
+    tearDown(() {
+      if (socket != null) {
+        socket.close();
+      }
+    });
+
+    test('should throw if connection cannot be established', () async {
+      expect(WSocket.connect(Uri.parse('ws://localhost:9999')),
+          throwsA(new isInstanceOf<WSocketException>()));
+    });
+
+    group('add()', () {
+      test('should be able to send a message', () async {
+        socket = await WSocket.connect(echoUri);
+        WSHelper helper = new WSHelper(socket);
+
+        socket.add('message');
+        await helper.messagesReceived(1);
+        expect(helper.messages.single, equals('message'));
+      });
+
+      test('should be able to send multiple messages', () async {
+        socket = await WSocket.connect(echoUri);
+        WSHelper helper = new WSHelper(socket);
+
+        socket.add('message1');
+        socket.add('message2');
+        await helper.messagesReceived(2);
+        expect(helper.messages, unorderedEquals(['message1', 'message2']));
+      });
+
+      test('should throw after sink has been closed', () async {
+        socket = await WSocket.connect(echoUri);
+        await socket.close();
+        expect(() {
+          socket.add('too late');
+        }, throwsStateError);
+      });
+    });
+
+    group('addError()', () {
+      // TODO: Adding an error forwards the error through to the underyling
+      // TODO:    WebSocket sink, which causes the error to be thrown, but
+      // TODO:    it can't be caught since it's thrown in an async zone.
+      // TODO:    Can we test this?
```

Comment:
Yeah I'll try to setup a reduced test case for this scenario and hit up the dart team if need be.